src/clusterfuzz/_internal/batch/service.py

```diff
@@ -0,0 +1,229 @@
+# Copyright 2025 Google LLC
+#
+# Licensed under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with
+# the License. You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Batch service.
+
+This module provides a high-level API for creating and managing remote tasks
+on GCP Batch. It abstracts away the details of the underlying batch client
+and provides a simple interface for scheduling ClusterFuzz tasks.
+"""
+import collections
+from typing import Dict
+from typing import List
+
+from clusterfuzz._internal.base import tasks
+from clusterfuzz._internal.base import utils
+from clusterfuzz._internal.base.tasks import task_utils
+from clusterfuzz._internal.batch.data_structures import BatchTask
+from clusterfuzz._internal.batch.data_structures import BatchWorkloadSpec
+from clusterfuzz._internal.batch.gcp import GcpBatchClient
+from clusterfuzz._internal.config import local_config
+from clusterfuzz._internal.datastore import data_types
+from clusterfuzz._internal.datastore import ndb_utils
+from clusterfuzz._internal.metrics import logs
+from clusterfuzz._internal.system import environment
+
+# See https://cloud.google.com/batch/quotas#job_limits
+MAX_CONCURRENT_VMS_PER_JOB = 1000
+
+
+def _get_batch_config():
+  """Returns the batch config. This function was made to make mocking easier."""
+  return local_config.BatchConfig()
+
+
+def is_remote_task(command: str, job_name: str) -> bool:
+  """Returns whether a task is configured to run remotely on GCP Batch.
+  
+  This is determined by checking if a valid batch workload specification can
+  be found for the given command and job type.
+  """
+  try:
+    _get_specs_from_config([BatchTask(command, job_name, None)])
+    return True
+  except ValueError:
+    return False
+
+
+def _get_config_names(batch_tasks: List[BatchTask]):
+  """"Gets the name of the configs for each batch_task. Returns a dict
+  that is indexed by command and job_type for efficient lookup."""
+  job_names = {task.job_type for task in batch_tasks}
+  query = data_types.Job.query(data_types.Job.name.IN(list(job_names)))
+  jobs = ndb_utils.get_all_from_query(query)
+  job_map = {job.name: job for job in jobs}
+  config_map = {}
+  for task in batch_tasks:
+    if task.job_type not in job_map:
+      logs.error(f"{task.job_type} doesn't exist.")
+      continue
+    if task.command == 'fuzz':
+      suffix = '-PREEMPTIBLE-UNPRIVILEGED'
+    else:
+      suffix = '-NONPREEMPTIBLE-UNPRIVILEGED'
```

Comment:
I wonder if this sort of thing will be shared with k8s.

Reply:
I don't think so. I think the best approach for Kubernetes is having the clusters as ON_DEMAND. Then even the machines can be a bit more expensive, the resources can be used better.